### PR TITLE
backend: (csl) Slightly rework printer to promote certain values to inline expressions

### PR DESCRIPTION
**Pull request by AntonLydike**
This patch makes the printer promote certain values to inlined expressions (e.g. not creating a new `const x = 5;` variable for an `arith.constant` operation, or for arithmetic expressions.

This is for now applied very carefully to a select few operations. Some checks are performed, such as that the value is not referenced at any point using the `AddressOf`operation, that the operation that produced the value is pure, and more.

### CODE REVIEW

File: tests/filecheck/backend/csl/print_csl.mlir [hunk 1]
```diff
@@ -449,10 +441,8 @@ csl.func @builtins() {
 // CHECK-NEXT: comptime {
 // CHECK-NEXT:   @export_symbol(args_no_return, "args_no_return");
 // CHECK-NEXT: }
-// CHECK-NEXT: const cst15 : i32 = 15;
-// CHECK-NEXT: const col : color = @get_color(cst15);
 // CHECK-NEXT: comptime {
-// CHECK-NEXT:   @rpc(@get_data_task_id(col));
+// CHECK-NEXT:   @rpc(@get_data_task_id(@get_color(15)));
```
Comment (AntonLydike) [inline tests/filecheck/backend/csl/print_csl.mlir:445]: This is a neat example of how this improves readabiltiy in the IR.

File: tests/filecheck/backend/csl/print_csl.mlir [hunk 2]
```diff
@@ -461,95 +451,67 @@ csl.func @builtins() {
...
-// CHECK-NEXT:     for(@range(i32, lb, ub, step)) |j| {
-// CHECK-NEXT:       const ix6 : i32 = i * ub;
-// CHECK-NEXT:       const ix6pj : i32 = ix6 +  j;
-// CHECK-NEXT:       const Axx : f32 = (A[ix6pj]) * (x[j]);
-// CHECK-NEXT:       tmp2 = tmp2 + Axx;
+// CHECK-NEXT:     for(@range(i32, 0, 6, 1)) |j| {
 // CHECK-NEXT:     }
-// CHECK-NEXT:     const tmp_plus_bi : f32 =  tmp2 + (b[i]);
-// CHECK-NEXT:     y[i] = tmp_plus_bi;
+// CHECK-NEXT:     y[i] = (tmp + (b[i]));
```
Comment (AntonLydike) [inline tests/filecheck/backend/csl/print_csl.mlir:483]: Something is wrong here :thinking:

Comment (AntonLydike) [reply tests/filecheck/backend/csl/print_csl.mlir]: Figured it out! It was the `scf.for` loop-carried variable trick from @dk949, should work now!

File: tests/filecheck/backend/csl/print_csl.mlir [hunk 3]
```diff
@@ -344,16 +344,15 @@ csl.func @builtins() {
 }) {sym_name = "layout.csl"} : () -> ()
 
 // CHECK-NEXT: // FILE: program.csl
-// CHECK-NEXT:
+// CHECK-NEXT:
```
Comment (tobiasgrosser) [inline tests/filecheck/backend/csl/print_csl.mlir:347]: Is this change intentional? Trailing whitespaces does not seem like a feature we would want.

**Review by n-io — APPROVED** [review]
Looks good, thanks for the work. I'd say fix trailing whitespace and merge
